cassandra-leaves-app/src/pages/index.js

```diff
@@ -58,85 +72,125 @@ export default function Home({ leavesData }) {
         setFilteredData(updatedData);
         setIsEditing(false);
         setFormState({ id: "", title: "", domain_name: "" });
+        setIsModalOpen(false);
+        setSuccessMessage("Record updated successfully!");
+        setTimeout(() => setSuccessMessage(""), 3000);
     };
 
-    // Delete a record
     const handleDelete = (id) => {
-        const updatedData = filteredData.filter((item) => item.id !== id);
+        setDeleteRecord(id);
+        setIsDeleteModalOpen(true);
+    };
+
+    const confirmDelete = () => {
+        const updatedData = filteredData.filter((item) => item.id !== deleteRecord);
         setFilteredData(updatedData);
+        setDeleteRecord(null);
+        setIsDeleteModalOpen(false);
         alert("Record deleted successfully.");
     };
 
-    // Pagination logic
+    const cancelDelete = () => {
+        setDeleteRecord(null);
+        setIsDeleteModalOpen(false);
+    };
+
     const totalPages = Math.ceil((filteredData?.length || 0) / itemsPerPage);
     const startIndex = (currentPage - 1) * itemsPerPage;
     const paginatedData = filteredData.slice(startIndex, startIndex + itemsPerPage);
 
+    const toggleDarkMode = () => {
+        setIsDarkMode((prevMode) => {
+            const newMode = !prevMode;
+            localStorage.setItem("darkMode", newMode);
+            return newMode;
+        });
+    };
+
```

Comment:
Nice job on implementing these handlers! They cover the functionality well. To make this code more organized, maintainable, and aligned with best practices, here are some suggestions:

### 1. **Organize Logic into Separate Custom Hooks**
Encapsulating related functionality into custom hooks can make the component more modular and reusable. For example:
- **Pagination Logic**: Extract pagination-related calculations (`totalPages`, `startIndex`, etc.) into a `usePagination` hook.
- **Dark Mode**: Extract `toggleDarkMode` into a `useDarkMode` hook that manages the state and localStorage logic.

Example:
```tsx
function usePagination(filteredData, itemsPerPage) {
    const totalPages = Math.ceil((filteredData?.length || 0) / itemsPerPage);
    const [currentPage, setCurrentPage] = useState(1);

    const paginatedData = useMemo(() => {
        const startIndex = (currentPage - 1) * itemsPerPage;
        return filteredData.slice(startIndex, startIndex + itemsPerPage);
    }, [filteredData, currentPage, itemsPerPage]);

    return { currentPage, setCurrentPage, totalPages, paginatedData };
}
```

---

### 2. **Separate Input and Form Management**
For `handleInputChange`, `handleAddRecord`, and `handleUpdateRecord`, consider creating a custom hook like `useFormState`. This ensures the form logic is reusable and isolated from the main component.

---

### 3. **Avoid Duplicating State**
`filteredData` is derived from `leavesData` and `searchQuery`, so it could be recalculated dynamically using `useMemo`:
```tsx
const filteredData = useMemo(() => {
    return leavesData.filter((item) =>
        item.title.toLowerCase().includes(searchQuery.toLowerCase())
    );
}, [leavesData, searchQuery]);
```
This removes the need to manage `filteredData` manually.

---

### 4. **Simplify Dark Mode Persistence**
The `toggleDarkMode` function is good, but for learning purposes, consider explaining how to abstract localStorage interaction into a utility:
```tsx
const [isDarkMode, setIsDarkMode] = useState(() =>
    JSON.parse(localStorage.getItem("darkMode") || "false")
);

const toggleDarkMode = () => {
    setIsDarkMode((prevMode) => {
        const newMode = !prevMode;
        localStorage.setItem("darkMode", JSON.stringify(newMode));
        return newMode;
    });
};
```

---

### 5. **Extract Modal Logic**
Modal handling (`isModalOpen`, `isDeleteModalOpen`, `handleDelete`, `cancelDelete`, etc.) can be extracted into a separate `useModal` hook to simplify the component.

---

### 6. **Add Form Validation**
Instead of using `alert` for form validation, consider providing inline error messages for better UX. This is a great learning opportunity to introduce validation patterns.

---

### 7. **Use Constants or Utilities for Messages**
To make the code more maintainable, centralize all success/error messages in a constants file or utility function:
```tsx
const messages = {
    addSuccess: "Record added successfully!",
    updateSuccess: "Record updated successfully!",
    deleteSuccess: "Record deleted successfully!",
};

setSuccessMessage(messages.addSuccess);
```

---

### 8. **Use Modern Practices**
Introduce modern practices like `async/await` for API interactions if applicable, and consider adding types (if using TypeScript) to make the code more robust and self-documenting.

---

By implementing these suggestions, the code will be easier to maintain and scale, and you’ll gain experience with patterns like custom hooks, derived state, and reusable logic. Let me know if you’d like to discuss any of these in more detail or need help implementing them!